feat: Add By Participant Expanded Views (M2-7921)

farmerpaul opened the pull request:
<!-- Use this template as a guide to describe your pull request, and adjust as necessary. -->
<!-- Include information that helps your peers review your updates and understand this    -->
<!-- repository's history of changes over time.                                           -->

<!-- Delete any options that are not relevant -->

- ~~[ ] Tests for the changes have been added~~ **(unit tests to be added in separate task)**
- [ ] Delivered the `fix` or `feature` branches into `develop` or `release` branches via `Squash and Merge` (to keep clean history)

### 📝 Description

<!-- Contributions are welcome! If there is a corresponding      -->
<!-- JIRA ticket, link to it by replacing `#` with ticket number -->

🔗 [Jira Ticket M2-7921](https://mindlogger.atlassian.net/browse/M2-7921)

This PR adds the expanded Subjects view to each Activity/Flow on the PDP's new By Participant tab. It includes these changes:

- Adding API support for the `/subjects/respondent/{subject_id}/activity-or-flow/{activity_or_flow_id}` endpoint, consumed by this expanded view
- Adding the `ExpandedView` sub-component of the `ByParticipant` screen component
- Adjusting the PDP's generic `ActivityListItem` component to be able to support custom expanded view content
- In the `ByParticipant` component, passing an instance of `ExpandedView` to each `ActivityListItem`, and handling fetching/refetching of data from the new endpoint to populate the `ExpandedView` and handle loading states gracefully

### 📸 Screenshots

https://github.com/user-attachments/assets/c7d0207c-7397-44da-842d-3cd97efea4d8

### 🪤 Peer Testing

> [!TIP]
> Ensure the `enableActivityAssign` feature flag is enabled in the environment you are testing from.

Review the AC of the [Jira ticket](https://mindlogger.atlassian.net/browse/M2-7921) to ensure behaviour of expanded view is in alignment. Specifically be sure to test this screen where:

- The Participant whose PDP you're on is manually assigned as a respondent to activities/flows about either themselves or other Participants
- The Participant whose PDP you're on is auto-assigned to certain activities/flows
- The Participant whose PDP you're on is **not** assigned to certain activities/flows, but has submitted answer data as a respondent in the past, about either themselves or other Participants
- Ensure that after creating assignments or removing assignments by clicking Assign/Unassign from any of the action menus on this screen causes data to be refreshed so that the screen is always up-to-date

farmerpaul committed:
feat: add API call to fetch subjects by respondent

farmerpaul committed:
fix: allow tables to have undefined sort props

Several instances of `DashboardTable` don't incorporate sorting and it
makes no sense for sorting props to be required by the component, so
made them optional. The expanded views of the By Participant tab will
also use `DashboardTable` without sorting.

farmerpaul committed:
feat: add By Participant `ExpandedView` component

This is the table that will appear inside a new collapsible area within
each `ActivityListItem` component (upcoming commit).

farmerpaul committed:
feat: adjust `getActionsMenu` for expanded views

farmerpaul committed:
feat: support expanded views in`ActivityListItem`

Make adjustments to `ActivityListItem` structure and styling to be able
to support a expanded view node optionally provided by the consumer.
If provided, then the item's `onClick` defaults to expanding/collapsing
the expanded view. Includes nuanced logic around loading states so that
expanded views can be refreshed, and stay open while being refreshed
without causing jarring UX.

farmerpaul committed:
feat: add `ExpandedView` to By Participant tab

In this commit we add logic to the By Participant tab to harness the
new support for expanded views in `ActivityListItem`, and pass an
instance of the `ExpandedView` component to each of those list items.
Data retrieval and refetching is managed in the `ByParticipant`
component, the same place where other API calls are already happening.

Each expanded view's data is stored as a property of a shared state
variable – a map of activities/flows ⇒ subject data. Another shared
state variable maps activities/flows ⇒ loading status, which is used
to display a spinner on the expand/collapse button for each
activity/flow.

farmerpaul committed:
chore: add translations

sultanofcardio commented on src/modules/Dashboard/features/Participant/Assignments/ByParticipant/ByParticipant.tsx:
```diff
@@ -1,5 +1,181 @@
+      if (!respondentSubject) return;
+
+      if (isExpanded) {
+        handleRefetchExpandedView(activityOrFlowId);
```
Adding `void` before the async function call indicates that we are intentionally ignoring the result/error
```suggestion
        void handleRefetchExpandedView(activityOrFlowId);
```

sultanofcardio replied:
I also think we should differentiate between refetching after the assign drawer closes, and before expanding the card. More often than not, the entries in this list will be the same after expanding the card but since we don't cache the list, it blocks the card from expanding.

Considering that, and the fact that this can be an expensive call depending on the number of results, I think we should cache the results of the call to `getTargetSubjectsByRespondentApi` for some time so that it doesn't perform a fetch on every expand. The cache can be ignored when fetching after the assign drawer closes

farmerpaul replied:
Since we don't have a pre-existing pattern for caching of endpoints in the Admin App (that I have been able to find, at least), rather than attempt to pull off a new caching system helter-skelter, I spoke to @mbanting and we agreed to split this performance optimization into a post-MI Assign launch fast-follow.

farmerpaul replied:
Created [M2-8051](https://mindlogger.atlassian.net/browse/M2-8051).

sultanofcardio commented on src/modules/Dashboard/features/Participant/Assignments/ByParticipant/ByParticipant.tsx:
```diff
@@ -1,5 +1,181 @@
+
+    // Refresh target subject data for any expanded views
+    Object.entries(expandedViewsData).forEach(([id, viewData]) => {
+      if (viewData) handleRefetchExpandedView(id);
```
```suggestion
      if (viewData) void handleRefetchExpandedView(id);
```

sultanofcardio reviewed (approved):
Looking good. I left a few comments for your consideration

farmerpaul committed:
fix: intentionally ignore async call return values

farmerpaul commented on src/modules/Dashboard/features/Participant/Assignments/ByParticipant/ByParticipant.tsx:
```diff
@@ -29,35 +40,86 @@ const ByParticipant = () => {
+        // If expanded view is closed, remove data to free up memory and minimize refetches
+        // (after delay to account for transition)
+        setTimeout(() => {
+          setExpandedViewsData((prev) => ({ ...prev, [activityOrFlowId]: undefined }));
```
Somehow your comment got lost @sultanofcardio, I guess during rebase? Just wanted to respond:

> Setting the entry to undefined will still keep it in the object, and later on Object.entries.forEach will still execute a re-fetch for it. Remove it instead

I accounted for `undefined` entries [above](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1943/files#diff-5fb2ade196f939ddef9b4ecea13b66a6e82119397515bbbb236193b54dd63cd5R75), only refetching if data is present for that entry. I considered your suggested approach and didn't like that the `const { …, ...rest } = …` approach took up 4 extra lines in the file, but it would mean one less conditional during refetch, so I'll do your version! (Also I shouldn't be so motivated by conciseness 😛)

sultanofcardio replied:
It didn't get lost, I later realised that it was accounted for so I deleted the comment. I should've probably replied to it instead 😅

farmerpaul replied:
I'm just as sneaky, rewording my reply to you like 5 times before I was satisfied lol… 🙌

farmerpaul committed:
fix: improve expanded views map logic

farmerpaul committed:
fix: abide by HTML police

Use Bootstrap's stretched-link CSS trick to make the primary button of
list items cover the entire container. Also add click slop to secdonary
buttons (i.e., the ⋯ menu).

farmerpaul committed:
fix: unmount `DataExportPopup` after use

This should prevent repeated downloads after first Export Data action.

farmerpaul committed:
fix: restore activity status tooltips

JustinNusca reviewed (approved):
Thanks for the effort in making sure the activity list items are accessible. 

In my original comment, I didn't account for one of the buttons having the same functionality as the original `onClick` pop did, so I think there was more wiggle room there than I realized.

Nested interactive elements can be difficult to maintain accessible UX for, and this appears to address everything I mentioned — even the off-hand comment about misclicks 😅.

Not required, but I wonder if a comment in the `ActivityListItem` component file that explicitly explains the `.primary-button` pattern as a means of adding click-ability to the component? The purpose isn't immediately obvious from just reading the CSS, and I can see reasoning for this being forgotten if this ever gets revisited much later.

farmerpaul commented:
Thanks for the review @JustinNusca! I did add a JSdoc [comment here](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1943/files#diff-3187bea639d37a9ccceccba3a217d38c0f5239e19267bd34490f3fdc038476c5R10-R15) to encourage consumers to use that class in the `children` prop, but it occurred to me now that people likely won't see that comment in their editor as they more likely would for other attribute-like props. Good call on adding it to the component definition itself. :+1:

farmerpaul committed:
doc: improve comments
